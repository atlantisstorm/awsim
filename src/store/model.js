import config from '../data/config.json';
import testInstances from '../data/running.instances.json';
import { extractInstances, getInstanceRawData, initFilters,  setSort } from '../utils';
const AWS  = require('aws-sdk');

export const defaultState = {
  loading: true,
  instances: [],
  selectedInstance: '',
  filteredInstances: [],
  filters: initFilters(),
  filtersOrdered : config.filters,
  fields: config.fields,
  modalSettings: {
    display: false,
    pageX: 5,
    pageY: 5,
    instance: {}
  },
  sortOptions: setSort('Name', 'asc'),
  aws: {
    accessKeyId: config.aws.accessKeyId,
    secretAccessKey: config.aws.secretAccessKey
  }
};

const filterInstances = (instances, filters) => {
  const filteredInstances = instances.filter(instance =>{
    let includeInstance = true;

    for (const key in filters) {
      const filter = filters[key];
      const filterField = filter.field;
      const filterValue = filter.currentValue;
      const filterMatch = filter.match;
      if (filterField === "Regions") {
        // The 'Regions' filter is only applied on loading/refresh request in useFetchData, so ignore here.
        continue;
      }

      if (filterMatch !== "regex") {
        if (filterValue && instance[key] !== filterValue) {
          includeInstance = false;
          break;
        }
      } else {
        // Currently just the "Name" field.
        const filterPattern = new RegExp('^' + filterValue);
        if (!filterPattern.test(instance[filterField])) {
          includeInstance = false;
          break;
        }
      }
    }

    return includeInstance;
  });

  return filteredInstances;
}

export const handleFetchData = ({ state, accessKeyId, secretAccessKey, region }) => {
  let instances = state.instances; // We do this to stop test data being over-written, bit hacky! :D
  if (process.env.NODE_ENV === "development") {
    instances = extractInstances(testInstances);
  } else {
    AWS.config.update({
      "accessKeyId": accessKeyId,
      "secretAccessKey": secretAccessKey,
      "region": region
    });

    const ec2 = new AWS.EC2({apiVersion: config.aws.apiVersion});
    const params = {
      DryRun: false
    };

    ec2.describeInstances(params, function(err, data) {
      if(err) {
        console.log("Error", err.stack);
      } else {
        instances = extractInstances(data);
      }
    });
  }

  return { ...state, loading: false, instances };
}

export const handleFilterInstances = ({ state }) => {
  const filteredInstances = filterInstances(state.instances, state.filters);
  return { ...state, filteredInstances };
}

export const handleFilterUpdate = ({ state, filterName, filterValue }) => {
  const filters = state.filters;
  const updatedFilter = { ...filters[filterName], currentValue: filterValue };
  const updatedFilters = { ...filters, [filterName]: updatedFilter };
  const filteredInstances = filterInstances(state.instances, updatedFilters);

  return { ...state, filteredInstances, filters: { ...filters, [filterName]: updatedFilter } };    
}

export const handleModalSetDisplay = ({ state, display }) => {
  const newModalSettings = {
    ...state.modalSettings,
    display: display
  };

  return { ...state, modalSettings: newModalSettings };
}

export const handleModalShowInstance = ({ state, instanceId, pageX, pageY}) => {
  const newModalSettings = {
    ...state.modalSettings,
    pageX,
    pageY,
    display: true,    
    instance: getInstanceRawData(state.instances, instanceId)
  };

  return { ...state, selectedInstance: instanceId, modalSettings: newModalSettings };
}

export const handleRefreshData = ({ state }) => {
  return { ...state, loading: true };
}

export const handleSortOptionsSet = ({ state, field, order }) => {
  return { ...state, sortOptions: setSort(field, order) };
}

export const Model = {
  handleFetchData,
  handleFilterInstances,
  handleFilterUpdate,
  handleModalSetDisplay,
  handleModalShowInstance,
  handleRefreshData,
  handleSortOptionsSet
};

export default Model;