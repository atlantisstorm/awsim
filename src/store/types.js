// Emmm... action types.
export const FETCH_DATA = "FETCH_DATA";
export const FILTER_INSTANCES = "FILTER_INSTANCES";
export const FILTER_UPDATE = "FILTER_UPDATE";
export const MODAL_SET_DISPLAY = "MODAL_SET_DISPLAY";
export const MODAL_SHOW_INSTANCE = "MODAL_SHOW_INSTANCE";
export const REFRESH_DATA = "REFRESH_DATA";
export const SORT_OPTIONS_SET = "SORT_OPTIONS_SET";