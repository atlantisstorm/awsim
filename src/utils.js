import config from './data/config.json';

export const compare = (fieldName, order) => {
  return function(a,b){
    const fieldValueA = a[fieldName].toLowerCase();
    const fieldValueB = b[fieldName].toLowerCase();
    
    let comparison = 0;
    if (fieldValueA > fieldValueB) {
      comparison = 1;
    } else if (fieldValueA < fieldValueB) {
      comparison = -1;
    }
    return order === "desc" ? (comparison * -1) : comparison;
  };
}
export const extractDynamicFilterValues = ({ filterField, instances }) => {
  let values = [];
  let filterSeen = {};
  for (let j in instances.sort(compare(filterField, "asc"))) {
    const fieldValue = instances[j][filterField];
    if (!filterSeen[fieldValue]) {
      values.push(fieldValue);
      filterSeen[fieldValue] = true;
    }
  }
  return values;
}

export const extractInstances = (data) => {
  const Instances = [];

  for (let i = 0; i < data.Reservations.length; i++) {
    const instances = data.Reservations[i].Instances;
    for (let j= 0; j < instances.length; j++ ) {
      const instData = instances[j];
      const Instance = {};

      Instance.Name = "";
      Instance.Environment = "";
      Instance.AutoScalingGroupName = "";

      for (let k = 0; k < instData.Tags.length; k++) {
        const Tag = instData.Tags[k];

        if (Tag.Key === "Name") {
          Instance.Name = Tag.Value;;
        }
        else if (Tag.Key === "Environment") {
          Instance.Environment = Tag.Value;
        }
        else if (Tag.Key === "aws:autoscaling:groupName") {
          Instance.AutoScalingGroupName = Tag.Value;
        }
      }

      Instance.StateName        = instData.State.Name || "";
      Instance.Platform         = instData.Platform || "";
      Instance.PublicIpAddress  = instData.PublicIpAddress || "";
      Instance.PrivateIpAddress = instData.PrivateIpAddress || "";
      Instance.InstanceId       = instData.InstanceId || "";
      Instance.ImageId          = instData.ImageId || "";
      Instance.InstanceType     = instData.InstanceType || "";

      // Convert launch time to yyyy-mm-dd hh:mm:ss format.
      Instance.LaunchTime       = instData.LaunchTime || "";
      const date = new Date(Instance.LaunchTime);
      Instance.LaunchTime = date.toLocaleString('en-GB', {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit',
        hour12: false
      }).replace(/(\d+)\/(\d+)\/(\d+),\s+(\d{2}:\d{2}:\d{2})/, '$3-$2-$1 $4');

      Instance._data = instData;
      Instances.push(Instance);
    }
  }

  return Instances;
}

// get original raw data.
export const getInstanceRawData = (instances, instance_id) => {
  let data = {};
  for (let idx in instances) {
    const instance = instances[idx];
    if (instance.InstanceId === instance_id) {
      data = instance._data;
      break;
    }
  }

  return data;
}

export const initFilters = () => {
  const filtersArray = config.filters;
  let filters = {};
  filtersArray.map((filter) => {
    filters[filter.field] = { ...filter, currentValue: "" };
  });
  return filters;
}

export const launchRemoteSession = (ip_address, targetPlatform) => {
  if (!(ip_address.length > 0)) {
    return false; // No ip address supplied, can happen when instance is not running.
  }

  let { cmd, errMsg } = buildRemoteSessionCmd(process.platform, targetPlatform, config.username, ip_address);

  if (errMsg) {
    alert(errMsg);
    return;
  }

  if (cmd.length > 0) {
    const child = require('child_process');
    child.exec(cmd, (err, stdout, stderr) => {
      if (err) {
        // node couldn't execute the command
        console.log(`node couldn't execute ${cmd}, err==`, err);
        return;
      }

      // the *entire* stdout and stderr (buffered)
      console.log(`stdout: ${stdout}`);
      console.log(`stderr: ${stderr}`);
    });
  }
}

export const buildRemoteSessionCmd = (thisPlatform, targetPlatform, username, ip_address) => {
  let cmd = "";
  let errMsg = "";
  if (!thisPlatform || !ip_address || !username) {
    errMsg = "You must supply all options!";
    return { errMsg };
  }

  if (!thisPlatform.match(/^(darwin|linux|win32)$/)) {
    errMsg = `Don't know how to build remote session for ${thisPlatform}`;
    return { errMsg };
  }

  if (!targetPlatform.match(/^(windows|)$/)) {
    errMsg = `Don't know how to build remote session for target platform ${targetPlatform}`;
    return { errMsg };
  }

  if (thisPlatform === "darwin") {
    const atc = `/tmp/awsim.terminal.command`;

    if (targetPlatform === 'windows') {
      const msRemoteDesktop = "\"/Applications/Microsoft\ Remote\ Desktop.app\"";
      const port = 3389;
      const address = `rdp://full address=s:${ip_address}:${port}`;
      const width = "desktopwidth:i:1024";
      const height = "desktopheight:i:768";
      const audio = "audiomode=i:0";
      const theme = "disable themes=i:1";
      const prompt = "prompt for credentials on client:i:0";
      cmd = `open -n -F -a ${msRemoteDesktop} "${address}&${audio}&${theme}&${width}&${height}&${prompt}"`;
    } else {
      cmd =
      `echo '#!/bin/bash' > ${atc};` +
      `echo 'ssh ${username}@${ip_address}' >> ${atc};` +
      `chmod 755 ${atc};` +
      `open -a Terminal -F ${atc}`;
    }
  }
  else if (thisPlatform === "linux") {
    const atc = `/tmp/awsim.terminal.command`;

    if (targetPlatform === 'windows') {
      // TODO
      errMsg = `Emmm..., sorry, no idea how to launch remote windows connection on ${platform}! :( )`;
    } else {
      cmd =
      `echo '#!/bin/bash' > ${atc};` +
      `echo 'ssh ${username}@${ip_address}' >> ${atc};` +
      `chmod 755 ${atc};` +
      `gnome-terminal -- ${atc}`;
    }
  }
  else if (thisPlatform === "win32") {
    if (targetPlatform === 'windows') {
      const width = 1024;
      const height = 768;
      cmd = `mstsc /v:${ip_address} /h:${height} /w:${width}`;
    }
    else {
      cmd = `start ssh ${username}@${ip_address}`;
    }
  }
  else {
    errMsg = `Emmm..., sorry, no idea how to launch remote connection on ${platform}! :( )`;
  }

  return { cmd, errMsg };
}

export const setSort = (fieldValue, orderValue) => {
  return {
    field: fieldValue,
    order: orderValue
  };
}