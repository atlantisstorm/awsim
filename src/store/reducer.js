import { useReducer } from 'react';
import config from '../data/config.json';
import testInstances from '../data/running.instances.json';
import { extractInstances, getInstanceRawData, initFilters,  setSort } from '../utils';
import {
  FETCH_DATA,
  FILTER_INSTANCES,
  FILTER_UPDATE,
  MODAL_SET_DISPLAY,
  MODAL_SHOW_INSTANCE,
  REFRESH_DATA,
  SORT_OPTIONS_SET
} from './types';

const AWS  = require('aws-sdk');

export const defaultState = {
  loading: true,
  instances: [],
  selectedInstance: '',
  filteredInstances: [],
  filters: initFilters(),
  filtersOrdered : config.filters,
  fields: config.fields,
  modalSettings: {
    display: false,
    pageX: 5,
    pageY: 5,
    instance: {}
  },
  sortOptions: setSort('Name', 'asc'),
  aws: {
    accessKeyId: config.aws.accessKeyId,
    secretAccessKey: config.aws.secretAccessKey
  }
};

const filterInstances = (instances, filters) => {
  const filteredInstances = instances.filter(instance =>{
    let includeInstance = true;

    for (const key in filters) {
      const filter = filters[key];
      const filterField = filter.field;
      const filterValue = filter.currentValue;
      const filterMatch = filter.match;
      if (filterField === "Regions") {
        // The 'Regions' filter is only applied on loading/refresh request in useFetchData, so ignore here.
        continue;
      }

      if (filterMatch !== "regex") {
        if (filterValue && instance[key] !== filterValue) {
          includeInstance = false;
          break;
        }
      } else {
        // Currently just the "Name" field.
        const filterPattern = new RegExp('^' + filterValue);
        if (!filterPattern.test(instance[filterField])) {
          includeInstance = false;
          break;
        }
      }
    }

    return includeInstance;
  });

  return filteredInstances;
}

export const reducer = (state, action) => {
  let payload;
  let instances;
  let filters;
  let newModalSettings;

  switch (action.type) {
    case FETCH_DATA: {
      payload = action.payload;
      instances = state.instances; // We do this to stop test data being over-written, bit hacky! :D

      if (process.env.NODE_ENV === "development") {
        instances = extractInstances(testInstances);
      } else {
        AWS.config.update({
          "accessKeyId": payload.accessKeyId,
          "secretAccessKey": payload.secretAccessKey,
          "region": payload.region
        });

        const ec2 = new AWS.EC2({apiVersion: config.aws.apiVersion});

        const params = {
          DryRun: false
        };

        ec2.describeInstances(params, function(err, data) {
          if(err) {
            console.log("Error", err.stack);
          } else {
            instances = extractInstances(data);
          }
        });
      }

      return { ...state, loading: false, instances };
    }

    case FILTER_INSTANCES: {
      const filteredInstances = filterInstances(state.instances, state.filters);
      return { ...state, filteredInstances };
    }

    case FILTER_UPDATE: {
      payload = action.payload;
      filters = state.filters;

      const filterName = payload.name;
      const filterValue = payload.value;
      const updatedFilter = { ...filters[filterName], currentValue: filterValue };
      const updatedFilters = { ...filters, [filterName]: updatedFilter };

      const filteredInstances = filterInstances(state.instances, updatedFilters);
      return { ...state, filteredInstances, filters: { ...filters, [filterName]: updatedFilter } };
    }

    case MODAL_SET_DISPLAY: {
      payload = action.payload;

      newModalSettings = {
        ...state.modalSettings,
        display: payload.display
      };
      return { ...state, modalSettings: newModalSettings };
    }

    case MODAL_SHOW_INSTANCE: {
      payload = action.payload;

      newModalSettings = {
        ...state.modalSettings,
        display: true,
        pageX: payload.pageX, 
        pageY: payload.pageY,
        instance: getInstanceRawData(state.instances, payload.instanceId)
      };
      return { ...state, selectedInstance: payload.instanceId, modalSettings: newModalSettings };
    }

    case REFRESH_DATA: {
      // This will trigger FETCH_DATA
      return { ...state, loading: true };
    }

    case SORT_OPTIONS_SET: {
      // This will trigger FETCH_DATA
      payload = action.payload;
      return { ...state, sortOptions: setSort(payload.field, payload.order) };
    }

    default: {
      return state;
    }
  }
};

export const initialiseReducer = (props) => {
  const iState = (props && props.initialState) ? props.initialState : defaultState;
  const [state, dispatch] = useReducer(reducer, iState);
  return { state, dispatch };
}