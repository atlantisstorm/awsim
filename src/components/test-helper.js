import React from 'react';
//import { render } from "@testing-library/react";
import Context from "../store/context";
import { initialiseReducer } from "../store/reducer";

export const RenderWithProvider = (props) => {
  const { children, testState, testDispatch }  = props;
  // Note: we pass any supplied testState to initialiseReducer rather than switching it like
  // we do for dispatch because of scoping.  Basically each rerender after a click/select
  // update event would mean we refresh again with the original data.
  const { state, dispatch: iDispatch } = initialiseReducer({ initialState: testState });
  const dispatch = testDispatch ? testDispatch : iDispatch;
  return (
    <Context.Provider value={{ state, dispatch }}>
      { children }
    </Context.Provider> 
  )
}