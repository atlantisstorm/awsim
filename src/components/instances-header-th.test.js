import React from 'react';
import { render, fireEvent } from "@testing-library/react";
import InstancesHeaderTh from './instances-header-th';
//import extractInstances from '../fixtures/extracted-instances';
import { defaultState } from '../store/reducer';
import { SORT_OPTIONS_SET } from '../store/types';
import { RenderWithProvider } from './test-helper';
import { extractDynamicFilterValues } from '../utils';

describe('InstancesHeaderTh component', () => {
  //const testState = {
  //  ...defaultState,
  //  loading: false,
  //  instances: extractInstances,
  //  filteredInstances: extractInstances
  //};

  const testState = defaultState;

  it('should match snapshot', () => {
    const fieldName = "Environment";

    const { asFragment } = render(
      <RenderWithProvider testState={testState}>
        <table>
          <thead>
            <tr>
              <InstancesHeaderTh fieldName={fieldName} />
            </tr>
          </thead>
        </table>
      </RenderWithProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call dispatch with expected field and order when header is clicked', () => {
    const dispatch = jest.fn();
    const fieldName = "Environment";

    const { getByTestId } = render(
      <RenderWithProvider testDispatch={dispatch} testState={testState}>
        <table>
          <thead>
            <tr>
              <InstancesHeaderTh fieldName={fieldName} />
            </tr>
          </thead>
        </table>
      </RenderWithProvider>
    );

    const headerSort = getByTestId('header-sort');
    fireEvent.click(headerSort);
    const payload = {
      field: "Environment",
      order: "asc"
    };
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({ type: SORT_OPTIONS_SET, payload });
  });

  it('should call dispatch with expected parameters when dynamic filter changes', () => {
    const fieldName = "Environment";

    const { container, getByTestId } = render(
      <RenderWithProvider testState={testState}>
        <table>
          <thead>
            <tr>
              <InstancesHeaderTh fieldName={fieldName} />
            </tr>
          </thead>
        </table>
      </RenderWithProvider>
    );

    const headerSort = getByTestId('header-sort');

    const ascIcon = '\u25B2';
    fireEvent.click(headerSort);
    expect(container.innerHTML).toContain(ascIcon);

    const descIcon = '\u25BC';
    fireEvent.click(headerSort);
    expect(container.innerHTML).toContain(descIcon);
  });
});