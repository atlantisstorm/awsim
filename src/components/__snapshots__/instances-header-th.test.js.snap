// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstancesHeaderTh component should match snapshot 1`] = `
<DocumentFragment>
  <table>
    <thead>
      <tr>
        <th
          class="{fieldName} {extra_class}"
          data-testid="header-sort"
        >
          <div
            class="title"
          >
            <div
              class="left"
            >
              Environment 
            </div>
            <div
              class="right"
            />
          </div>
        </th>
      </tr>
    </thead>
  </table>
</DocumentFragment>
`;
